crates/krator/src/store.rs

```diff
@@ -0,0 +1,136 @@
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use kube::api::DynamicObject;
+use kube::api::GroupVersionKind;
+use serde::de::DeserializeOwned;
+use tokio::sync::RwLock;
+
+use crate::object::ObjectKey;
+
+type ResourceMap = HashMap<GroupVersionKind, HashMap<ObjectKey, serde_json::Value>>;
+
+/// Defines Store type for caching Kubernetes objects locally.
+///
+/// * State is held in `Arc` so it is cheap to clone.
+/// * Collections are scoped by {group, version, kind, namespace, name}.
+/// * Objects are stored as [DynamicObject](kube::api::DynamicObject)s.
+///
+/// ```
+/// # use krator::Store;
+/// # use k8s_openapi::api::core::v1::Pod;
+/// #
+/// # #[tokio::main]
+/// # async fn main() -> anyhow::Result<()> {
+/// let store = Store::new();
+///
+/// let pod = Pod::default();
+///
+/// store.insert(Some("namespace"), "name", pod).await;
+///
+/// // Elsewhere
+/// let pod = store.get::<Pod>(Some("namespace"), "name").await?;
+///
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Clone)]
+pub struct Store {
+    objects: Arc<RwLock<ResourceMap>>,
+}
+
+impl Default for Store {
+    fn default() -> Self {
+        Store::new()
+    }
+}
+
+impl Store {
+    /// Initialize empty store.
+    pub fn new() -> Self {
+        Store {
+            objects: Arc::new(RwLock::new(HashMap::new())),
+        }
+    }
+
+    /// Insert a new object.
+    pub async fn insert<R: 'static + k8s_openapi::Resource + Sync + Send + serde::Serialize>(
+        &self,
+        namespace: Option<&str>,
+        name: &str,
+        object: R,
+    ) {
+        let mut objects = self.objects.write().await;
+        let key = GroupVersionKind::gvk(R::GROUP, R::VERSION, R::KIND).unwrap();
+        let resource_objects = (*objects).entry(key).or_insert_with(HashMap::new);
+        let object_key = ObjectKey::new(namespace.map(|s| s.to_string()), name.to_string());
+        resource_objects.insert(object_key, serde_json::to_value(&object).unwrap());
```

Comment:
Just to clarify, are you storing these as JSON values to avoid having to use `Box<dyn Any>`? Are there going to be any performance problems to worry about?

Reply:
I was originally planning to store a `Box<dyn Any>`, but there is kind of a convoluted reason this won't work:

1. I start with `Watch` which just has the `GroupVersionKind` information for the resource and not the actual type. 
2. I construct a `kube_runtime::Watcher` using this which yeilds `DynamicObject`.
3. `DynamicObject` stores the data as `serde_json::Value`. 
4. I won't have the type information again to convert back until the user queries it. 

It seemed reasonable to use `serde_json::Value` since that is how `kube` appears to be trending for storing type-erased objects and we are tightly integrated with that crate. I'm not totally sure about performance but I definitely dont like the serialization and de serialization that happens when the user fetches an object. 

There may also be some changes in the pipeline for the `kube` crate that will simplify certain things.